ExtraDry/ExtraDry.UploadTools/UploadTools.cs

```diff
@@ -0,0 +1,204 @@
+                    "py", "go", "app", "scpt", "scptd" // Additional filetypes.
+                });
+                ExtensionOnlyBlacklist.AddRange(FileTypeBlacklist);
+                ExtensionOnlyBlacklist.AddRange(new List<string>() { "apk", "jar" });
+            }
+        }
+
+        /// <summary>
+        /// Given a filename, this will:
+        ///  - Replace spaces with hyphens
+        ///  - Remove invalid characters and replace with hyphens. 
+        ///  - Ensure the file name begins with a valid character
+        ///  - Trim invalid characters from start and end from filename
+        /// </summary>
+        public static string CleanFilename(string filename)
+        {
+            // Sort out the filename part
+            var fileOnly = Path.GetFileNameWithoutExtension(filename);
+
+            // replace all the invalid characters
+            var cleanedFilename = Regex.Replace(fileOnly, cleaningRegex, "-"); // Replace invalid characters with hyphens
+
+            // collapse duplicate hyphens and underscores
+            cleanedFilename = Regex.Replace(cleanedFilename, @"-+", "-"); // Collapse duplicate hyphens
+            cleanedFilename = Regex.Replace(cleanedFilename, @"_+", "_"); // Collapse duplicate underscores
+            cleanedFilename = Regex.Replace(cleanedFilename, @"[-\.]{2,}", "."); // Collapse -.- combinations to a .
+
+            // Remove trailing or leading hyphens
+            cleanedFilename = cleanedFilename.Trim('-');
+            cleanedFilename = cleanedFilename.TrimEnd('_');
+
+            // Now the file extension, the only extensions that aren't alphanumeric are files that we likely should be rejecting
+            // The addition of the hyphen as well allows for wordpress content and other similar proprietary but possible files.
+            // https://www.file-extensions.org/extensions/begin-with-special-characters
+            // https://www.file-extensions.org/filetype/extension/name/miscellaneous-files
+            var extension = Path.GetExtension(filename).Trim('.');
+            extension = Regex.Replace(extension, @"[^a-zA-Z0-9\-]", string.Empty);
+
+            cleanedFilename = $"{cleanedFilename}.{extension}";
+            return cleanedFilename;
+        }
+
+        /// <summary>
+        /// Determines whether a file can be uploaded, by using the following rules
+        /// - The filename must not contain no special characters
+        /// - The filename must not not have an extension that is in our blacklist
+        /// - The filename must be included in our whitelist
+        /// - The magic bytes, filename and mimetype must match if present
+        /// - If the file type is of a known xml type, it must not contain a script tag
+        /// If all of these are true, then true is returned. Else, a <see cref="DryException"/> with details is thrown
+        /// </summary>
+        public static bool CanUpload(string filename, string mimetype, byte[] content)
```

Comment:
This contract never returns false, it throws exceptions on all errors.  It feels like too much is being crammed into a method and that limits my ability to get custom behaviours as a consumer.  E.g. I can't simply write `if(!CanUpload()) { Ignore(); }`  Maybe the DataValidator class is a better pattern?  E.g. `var validator = new FileValidator(file)`, then I can do `validator.Success` or `validator.ThrowIfError()`...

Reply:
That's a good idea. When I changed the false returns to throw and provide details, I didn't like it, but I couldn't think of anything better. I was too close to it I think. I'll look at that refactor today

Reply:
I've made this change. Added a new class called FileChecker which does just this. The constructor takes the file details, and it runs the validations in there. Then you can check CanUpload or ThrowIfError for the results.